### PR TITLE
Dialog refactor

### PR DESCRIPTION
**Pull request by blink1073**
Refactored dialog to properly use widget lifecycle events.  Extracts the value from the body widget before disposing of it.  Also accepts virtual dom content for static dialogs.

### CODE REVIEW

File: packages/application-extension/src/index.ts [hunk 1]
```diff
@@ -47,12 +50,17 @@ const mainPlugin: JupyterLabPlugin<void> = {
...
+        h.p([
+          'JupyterLab build is out of date',
+          h.br(),
+          'Please run',
+          h.code('jupyter lab build'),
+          'from',
+          h.br(),
+          'the command line and relaunch'
+        ])
+      ]);
```
Comment (sccolbert) [inline packages/application-extension/src/index.ts]: The arrays are not needed. The `h` functions accept varargs.

File: packages/apputils/src/clientsession.ts [hunk 2]
```diff
@@ -866,6 +872,43 @@ namespace Private {
    */
   export
   function selectKernel(session: ClientSession): Promise<Kernel.IModel> {
+    let select = Dialog.okButton({ label: 'SELECT' });
```
Comment (sccolbert) [inline packages/apputils/src/clientsession.ts]: why is this button a separate line, but not the cancel button?

File: packages/apputils/src/dialog.ts [hunk 3]
```diff
@@ -34,26 +34,23 @@ import {
...
   /**
    * Create a dialog panel instance.
    *
    * @param options - The dialog setup options.
    */
-  constructor(options: Dialog.IOptions={}) {
+  constructor(options: Dialog.IOptions<T>={}) {
     super();
     this.addClass('jp-Dialog');
     options = Private.handleOptions(options);
```
Comment (sccolbert) [inline packages/apputils/src/dialog.ts]: This form of option handling is not `null` safe. The type of `options` is still an object with all optional properties.

File: packages/docmanager/src/dialogs.ts [hunk 4]
```diff
@@ -57,7 +58,20 @@ interface IFileContainer extends JSONObject {
  */
 export
 function renameDialog(manager: IDocumentManager, oldPath: string): Promise<Contents.IModel | null> {
-  return (new RenameHandler(manager, oldPath)).showDialog();
+  let body = new RenameHandler(oldPath);
+  return showDialog({
+    title: 'Rename File',
+    body,
```
Comment (sccolbert) [inline packages/docmanager/src/dialogs.ts]: `body: new RenameHandler(oldPath)`

File: packages/notebook/src/actions.ts [hunk 5]
```diff
@@ -36,7 +40,16 @@ import {
...
+const TRUST_MESSAGE = h.p([
+  'A trusted Jupyter notebook may execute hidden malicious code when you ',
+  'open it.',
+  h.br(),
+  'Selecting trust will re-render this notebook in a trusted state.',
+  h.br(),
+  'For more information, see the',
+  h.a({ href: 'http://ipython.org/ipython-doc/2/notebook/security.html' },
+      'Jupyter security documentation'),
+]);
```
Comment (sccolbert) [inline packages/notebook/src/actions.ts]: array not needed.

File: packages/apputils/src/clientsession.ts [hunk 6]
```diff
@@ -866,6 +872,43 @@ namespace Private {
...
+    return showDialog({
+      title: 'Select Kernel',
+      body: new KernelSelector(session),
+      buttons: [Dialog.cancelButton(), select]
+    }).then(result => {
+      if (!result.accept) {
+        return void 0;
```
Comment (sccolbert) [inline packages/apputils/src/clientsession.ts]: the return type should be `Promise<Kernel.IModel | null>` and then you can just `return result.value`, or this `undefined` needs to be replaced with an actual model.

File: packages/apputils/src/dialog.ts [hunk 7]
```diff
@@ -65,21 +62,29 @@ class Dialog extends Widget {
...
     let content = new Panel();
     content.addClass('jp-Dialog-content');
     layout.addWidget(content);
 
+    this._body = options.body;
+
     let header = renderer.createHeader(options.title);
     let body = renderer.createBody(options.body);
     let footer = renderer.createFooter(this._buttonNodes);
     content.addWidget(header);
```
Comment (sccolbert) [inline packages/apputils/src/dialog.ts]: why is there an extra content widget, which is the only child of the dialog? Why not just add the header, body, and footer directly to the layout?

File: packages/apputils/src/dialog.ts [hunk 8]
```diff
@@ -65,21 +62,29 @@ class Dialog extends Widget {
...
     content.addWidget(body);
     content.addWidget(footer);
+
+    this._primary = this._buttonNodes[this._defaultButton];
+
+    if (options.primaryElement) {
```
Comment (sccolbert) [inline packages/apputils/src/dialog.ts]: Not a fan of "primaryElement" name, when its actually a query string. Something more descriptive like "focusNodeSelector" or "focusQuerySelector" or "primaryElementSelector" would be better.

File: packages/apputils/src/dialog.ts [hunk 9]
```diff
@@ -281,23 +286,33 @@ class Dialog extends Widget {
...
-    promise.resolve(item);
+    let body = this._body;
+    let value: T | null = null;
+    if (body instanceof Widget && typeof body.getValue === 'function') {
+      value = body.getValue();
+    }
+    this.dispose();
+    promise.resolve({
+      button,
+      accept: button.accept,
```
Comment (sccolbert) [inline packages/apputils/src/dialog.ts]: this seems a bit redundant. If accept always comes from the button, why include it as a separate property?

File: packages/apputils/src/dialog.ts [hunk 10]
```diff
@@ -34,29 +34,26 @@ import {
...
    * Create a dialog panel instance.
    *
    * @param options - The dialog setup options.
    */
-  constructor(options: Dialog.IOptions={}) {
+  constructor(options: Dialog.IOptions<T>={}) {
     super();
     this.addClass('jp-Dialog');
-    options = Private.handleOptions(options);
+    options = Private.handleOptions(options || {});
```
Comment (sccolbert) [inline packages/apputils/src/dialog.ts]: why the `|| {}`?

File: packages/apputils/src/dialog.ts (hunk 7 above)
Comment (blink1073) [inline packages/apputils/src/dialog.ts]: The content is used to center the dialog on the host node with a faded background.

Comment (sccolbert) [inline packages/apputils/src/dialog.ts]: got it.

**Comment by blink1073** [comment]
Rebased and adding strict null checks.

**Pull request merged**